scripts/test_usb_scan.py

```diff
@@ -0,0 +1,393 @@
+    def sys_name(self):
+        return self.d.get("name")
+
+    def get_prop(self):
+        return self.d.get("props")
+
+    def get_attr(self):
+        return self.d.get("attrs")
+
+    @property
+    def attributes(self):
+        return MocDeviceAttrs(self)
+
+    def __iter__(self):
+        for name in self.get_prop():
+            yield name
+
+    def __len__(self):
+        return len(self.get_prop())
+
+    def __getitem__(self, name):
+        return self.get_prop().get(name)
+
+
+class MocEnumerator(object):
+
+    def __init__(self, ds):
+        self.ds = ds
+
+    def __iter__(self):
+        for d in self.ds:
+            yield MocDevice(d)
+
+
+class MocContext(object):
+
+    def __init__(self, devices, interfaces):
+        self.devices = devices
+        self.interfaces = interfaces
+
+    def list_devices(self, **kwargs):
+        if "usb" == kwargs.pop("subsystem"):
+            dev_type = kwargs.pop("DEVTYPE")
+            if "usb_device" == dev_type:
+                return MocEnumerator(self.devices)
+            elif "usb_interface" == dev_type:
+                return MocEnumerator(self.interfaces)
+        return MocEnumerator([])
+
+
+def mock_setup(mod, devices, interfaces, path):
+    mod.log.error = test_log
+    mod.log.debug = test_log
+    mod.Policy._PATH = path
+    mod.pyudev.Context = mock.Mock(return_value=MocContext(
+        devices, interfaces))
+
+
+@nottest
+def test_log(m):
+    global log_file
+    with open(log_file, "a+") as f:
+        f.write("{}: {}\n".format(sys.argv[0], m))
+
+
+class TestUsbScan(unittest.TestCase):
+
+    def setUp(self):
+        global log_file
+        try:
+            self.work_dir = tempfile.mkdtemp(prefix="test_usb_scan")
+            log_file = os.path.join(self.work_dir, "user.log")
+        except:
+            raise
+
+    def tearDown(self):
+        shutil.rmtree(self.work_dir, ignore_errors=True)
+
+    @nottest
+    def test_usb_common(self, moc_devices, moc_interfaces, moc_results,
+                        path="./scripts/usb-policy.conf"):
+        import usb_scan
+        mock_setup(usb_scan, moc_devices, moc_interfaces, path)
+
+        devices, interfaces = usb_scan.get_usb_info()
+
+        # debug info
+        usb_scan.log_list(devices)
+        usb_scan.log_list(interfaces)
+
+        # match interface to device
+        for i in interfaces:
+            for d in devices:
+                if i.is_child_of(d):
+                    d.add_interface(i)
+
+        usb_scan.log_list(devices)
+
+        # do policy check
+        policy = usb_scan.Policy()
+        pusbs = [usb_scan.to_pusb(d) for d in devices if
+                 d.is_ready() and policy.check(d)]
+
+        # pass pusbs in json to XAPI
+        self.assertItemsEqual(pusbs, moc_results)
+
+    @nottest
+    def test_usb_exit(self, moc_devices, moc_interfaces, moc_results,
+                        path="./scripts/usb-policy.conf"):
+        with self.assertRaises(SystemExit):
```

Comment:
I think here you should also check that the failure did show the correct log error, at least the immutable part of it.
You could use `@raises(SystemExit)` as attribute and `assert` over the initial part of the error message for example

Reply:
OK.
    @nottest
    def test_usb_exit(self, devices, interfaces, results,
                      path="./scripts/usb-policy.conf", msg=""):
        with self.assertRaises(SystemExit) as cm:
            self.test_usb_common(devices, interfaces, results, path)
        exception = cm.exception
        print exception.message
        if msg:
            self.assertIn(msg, exception.message)